pkg/kube/controllers/waitservice/pod_mutator.go

```diff
@@ -0,0 +1,129 @@
+package waitservice
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"code.cloudfoundry.org/quarks-operator/pkg/kube/apis"
+	"code.cloudfoundry.org/quarks-operator/pkg/kube/util/operatorimage"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"code.cloudfoundry.org/quarks-utils/pkg/config"
+)
+
+var (
+	// WaitKey is the key for identifying which service the pod has to wait for
+	WaitKey = fmt.Sprintf("%s/wait-for", apis.GroupName)
+)
+
+// PodMutator for adding waiting init containers for linked services
+type PodMutator struct {
+	client  client.Client
+	log     *zap.SugaredLogger
+	config  *config.Config
+	decoder *admission.Decoder
+}
+
+// Check that PodMutator implements the admission.Handler interface
+var _ admission.Handler = &PodMutator{}
+
+// NewPodMutator returns a new mutator that adds waiting InitContainers
+func NewPodMutator(log *zap.SugaredLogger, config *config.Config) admission.Handler {
+	mutatorLog := log.Named("waitservice-pod-mutator")
+	mutatorLog.Info("Creating a Pod mutator for WaitService")
+
+	return &PodMutator{
+		log:    mutatorLog,
+		config: config,
+	}
+}
+
+func validWait(annotations map[string]string) bool {
+	valid := false
+	if annotations[WaitKey] != "" {
+		valid = true
+	}
+
+	return valid
+}
+
+// Handle checks if the pod has the "wait-for" annotation and injects an initcontainer waiting for the service
+func (m *PodMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+	err := m.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	updatedPod := pod.DeepCopy()
+	if validWait(pod.GetAnnotations()) {
+		err = m.addInitContainer(ctx, updatedPod)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+	}
+
+	marshaledPod, err := json.Marshal(updatedPod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func (m *PodMutator) addInitContainer(ctx context.Context, pod *corev1.Pod) error {
+	annotations := pod.GetAnnotations()
+	serviceName, ok := annotations[WaitKey]
+	if !ok {
+		return fmt.Errorf("no annotations %s found", WaitKey)
+	}
+
+	if serviceName == "" {
+		return fmt.Errorf("service name in '%s' empty", WaitKey)
+	}
+
+	pod.Spec.InitContainers = append(pod.Spec.InitContainers, createWaitContainer(&serviceName)...)
```

Comment:
Why does createWaitContainer return an array of containers (with one element)?
Is this meant to be a prepend instead of append? `append(createWaitContainer(), InitContainers...)`  Would make sense I guess, as we wouldn't run other init containers before the wait condition is fulfilled?

Reply:
I was hoping to re-use it in case we would require "waiting for more services".

About the append I'm not sure, I thought the requirement was for the container itself, not for the initcontainers of the pod... but it makes sense, will change it to a pretend :+1:

Reply:
fixed in https://github.com/cloudfoundry-incubator/quarks-operator/pull/921/commits/4be4974b0526b89a3715709b5329d00f5bba2334 (changed the function to accept more services) and https://github.com/cloudfoundry-incubator/quarks-operator/pull/921/commits/4013406677d619df8b2d25f7ca73027a15796539 ! thanks :slightly_smiling_face: